File: Baekjoon/DFS와 BFS/7576 토마토/Main(1).java
// BFS 문제
// BFS로 탐색하면서, 전날에 삽입한 위치들을 모두 큐에서 뽑았을 경우 하루가 지난걸로 친다.

// 근데, 다른 풀이 보니 그냥 map[next.y][next.x] = map[cur.y][cur.x] + 1로 만들어서 max 값을 출력하는게 더 괜찮은 방법이다.

import java.io.*;
import java.util.*;

public class Main {
    static int[][] map;
    static boolean[][] visited;
    static int[] x_move = {1, 0, -1, 0};
    static int[] y_move = {0, -1, 0, 1};
    static int answer = -1;

    static int cnt_inserted = 0;

    static Queue<Pos> queue = new LinkedList<>();


    static class Pos {
        int y, x;

        Pos(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }

    public static void solution(int N, int M) {
        int cnt_poll = 0;
        int cnt_tmp = 0;

        while (!queue.isEmpty()) {
            Pos cur = queue.poll();
            cnt_poll++;

            for (int i = 0; i < 4; i++) {
                Pos next = new Pos(cur.y + y_move[i], cur.x + x_move[i]);

                if (next.y < 0 || next.y >= N || next.x < 0 || next.x >= M || visited[next.y][next.x]) continue;

                if (map[next.y][next.x] == 0) {
                    visited[next.y][next.x] = true;
                    map[next.y][next.x] = 1;
                    queue.offer(next);
                    cnt_tmp++;
                }
            }

            if (cnt_poll == cnt_inserted) {
                answer++;
                cnt_inserted = cnt_tmp;
                cnt_tmp = 0;
                cnt_poll = 0;
            }
        }

        boolean flag = false;

        for (int[] y : map) {
            for (int x : y) {
                if (x == 0) {
                    flag = true;
                    break;
                }
            }
        }

        if (flag) answer = -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        map = new int[N][M];
        visited = new boolean[N][M];

        boolean flag = false;

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int k = 0; k < M; k++) {
                int tmp = Integer.parseInt(st.nextToken());
                if (tmp == 1) {
                    cnt_inserted++;
                    queue.offer(new Pos(i, k));
                } else if (tmp == 0) flag = true;
                map[i][k] = tmp;
            }
        }

        if (!flag) {
            answer = 0;
            System.out.println(answer);
            return;
        }

        solution(N, M);
        System.out.println(answer);
    }
}

File: Baekjoon/DFS와 BFS/13549 숨바꼭질 3/Main (1).kt
// BFS 문제
// 오래걸림

// 처음엔 DP 인줄 알았는데, 메모리가 안됨
// 이건 나중에서야 알았는데,
// 순간이동 > 뒤 > 앞 순 (가중치가 낮은 순)으로 큐에 삽입하면 답을 보장한다.
// 이는, 가중치가 0 또는 1인 그래프의 최단거리를 구하는 것과 같아서 0-1 BFS라고 한다.

import java.util.LinkedList

val MAX = 1_000_000
val move = intArrayOf(1, -1, 2)
val depth = IntArray(MAX) { 1e9.toInt() }

data class Pos(var value: Int, var depth: Int)

fun solution(N: Int, M: Int): Int {
    var tmp = Pos(N, 0)
    val queue = LinkedList<Pos>()

    if (N > M) return N - M
    if (N == M) return 0

    if (N == 0) {
        tmp.value += 1
        tmp.depth += 1
        depth[1] = 1
        depth[0] = 0
        queue.offer(Pos(tmp.value, tmp.depth))
    } else {
        queue.offer(Pos(tmp.value, tmp.depth))
        depth[tmp.value] = tmp.depth
    }

    while (tmp.value * 2 <= M) {
        tmp.value *= 2
        depth[tmp.value] = tmp.depth
        queue.offer(Pos(tmp.value, tmp.depth))
    }

    while (!queue.isEmpty()) {
        var cur = queue.poll()

        if (cur.value == M) {
            return depth[cur.value]
        }

        for (i in 0..2) {
            if (i == 2) {
                val next = Pos(cur.value * move[i], cur.depth)

                if (next.value >= MAX - 1 || next.value < 0 || depth[next.value] <= next.depth) continue

                queue.offer(next)
                depth[next.value] = next.depth
                continue
            }

            val next = Pos(cur.value + move[i], cur.depth + 1)

            if (next.value >= MAX - 1 || next.value < 0 || depth[next.value] <= next.depth) continue

            queue.offer(next)
            depth[next.value] = next.depth
        }
    }

    return -1
}

fun main() {
    val (N, M) = readln().split(" ").map { it.toInt() }

    println(solution(N, M))
}